### PR TITLE
DbKvs sweep refactors to address last PR comments

### PR DESCRIPTION
**Pull request by gbonik**
As dxiao pointed out, CandidatePagingState does too many things. The
solution is to make CellTsPairLoader return an iterator and move all
paging logic inside that iterator, and remove CellTsPairLoaderFactory.
Then we replace CandidatePagingState with CandidateGroupingIterator,
which now doesn't have any paging logic and only takes care of grouping
(cell, ts) pairs by cell.

Also remove CandidatePageJoiningIterator because it's not used anymore.
We could have done this earlier but I forgot.

Plan for the future:
- Since we gave up on the in-database filtering idea, we can replace
  KVS.getCandidateCellsForSweeping() with a simpler call like
  getCellTsPairs() which would do what CellTsPairLoader does now.
- Implement the new call for the remaining KVS's. We need at least
  the InMemoryKVS. We should decide the destiny of JdbcKVS and CqlKVS.
- Remove all remaining usages of getRangeOfTimestamps(). I think that's
  basically deleteRange() on Cassandra.
- Remove KVS.getRangeOfTimestamps()!

[no release notes]

**Priority (whenever / two weeks / yesterday)**:
Ideally before we cut for the LTS

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2497)

### CODE REVIEW

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java [hunk 1]
```diff
@@ -199,6 +219,51 @@ private void appendEmptyValueFlagSelector(FullQuery.Builder builder) {
...
+        private void computeNextStartPosition(List<CellTsPairInfo> results, boolean scannedSingleRow) {
+            if (results.size() < sqlRowLimit) {
+                if (scannedSingleRow) {
+                    // If we scanned a single row and reached the end, we just restart
+                    // from the lexicographically next row
+                    byte[] nextRow = RangeRequests.getNextStartRowUnlessTerminal(false, startRowInclusive);
+                    if (nextRow == null) {
+                        reachedEnd = true;
+                    } else {
+                        startRowInclusive = nextRow;
```
[the remark this answers is not in the document]
Comment (gsheasby) [reply atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java]: Maybe not `cells..currentRow` as only Oracle cares about this, and the Token could be used across Postgres and Oracle (and maybe Cassandra?)

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/DbKvsGetCandidateCellsForSweeping.java [hunk 2]
```diff
@@ -18,61 +18,26 @@
...
 import com.google.common.collect.Iterators;
 import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweeping;
 import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweepingRequest;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
-import com.palantir.atlasdb.keyvalue.dbkvs.impl.sweep.CandidatePagingState.StartingPosition;
 
 public class DbKvsGetCandidateCellsForSweeping {
 
-    private final CellTsPairLoaderFactory cellTsPairLoaderFactory;
+    private final CellTsPairLoader cellTsPairLoaderFactory;
```
Comment (gsheasby) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/DbKvsGetCandidateCellsForSweeping.java]: Found another non-rename. Intellij lets you do the rename automagically.

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/DbKvsGetCandidateCellsForSweeping.java [hunk 3]
```diff
@@ -18,61 +18,26 @@
...
 
     public Iterator<List<CandidateCellForSweeping>> getCandidateCellsForSweeping(
             TableReference tableRef,
             CandidateCellForSweepingRequest request) {
-        CellTsPairLoader loader = cellTsPairLoaderFactory.createCellTsLoader(tableRef, request);
-        CandidatePagingState state = CandidatePagingState.create(request.startRowInclusive());
-        return Iterators.filter(new PageIterator(loader, state), page -> !page.isEmpty());
+        Iterator<List<CellTsPairInfo>> cellTsIter = cellTsPairLoaderFactory.createPageIterator(tableRef, request);
+        Iterator<List<CandidateCellForSweeping>> rawIter = CandidateGroupingIterator.create(cellTsIter);
+        return Iterators.filter(rawIter, page -> !page.isEmpty());
```
Comment (gsheasby) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/DbKvsGetCandidateCellsForSweeping.java]: Does this Iterator do an eager page load of the kind we were trying to avoid in `OracleCellTsPageLoader.PageIterator`?

**Comment by gsheasby** [comment]
Thanks for following through quickly with the refactor!

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CellTsPairToken.java, line 23 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfIBW0TI_S857oqvx-:-KwfIBW0TI_S857oqvx0:bt492vz) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CellTsPairToken.java#L23)):*
> ```Java
> import com.palantir.atlasdb.encoding.PtBytes;
> 
> public final class CellTsPairToken {
> ```

Value.Immutable?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CellTsPairToken.java, line 23 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfIBW0TI_S857oqvx-:-KwfIKVZLY_rYwIFocn7:b-r1bqwy) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CellTsPairToken.java#L23)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Value.Immutable?
</blockquote></details>

We'll have a builder that will make the various types of tokens easier to construct..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java, line 116 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfIiCxTYqozvxGGwjj:-KwfIiCxTYqozvxGGwjk:b-qsyuya) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java#L116)):*
> ```Java
>         final int sqlRowLimit;
> 
>         CellTsPairToken token;
> ```

final?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java, line 230 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwcDeN6e96Tb6r5Fenl-r1-230:-KwfJckES014KvABrC_H:b-u25ssb) ([raw file](https://github.com/palantir/atlasdb/blob/95fb4dd20d513a23f8c00f28dd905a112c73a49a/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java#L230)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

We're iterating through the results twice - once to bundle each one into a `CellTsPairInfo`, and again to figure out the count of `cellTsPairsAlreadyExaminedInCurrentRow`.

Is it worth cutting this iteration time in half by updating `cellTsPairsAlreadyExaminedInCurrentRow` within `loadPage()`?
</blockquote></details>

Discussed offline with Glenn, doesnt seem like this significantly effects the time complexity.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java, line 258 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfKJgHvUBaPh65Thw7:-KwfKJgHvUBaPh65Thw8:b-91mfv2) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java#L258)):*
> ```Java
>             } else {
>                 CellTsPairInfo lastResult = Iterables.getLast(results);
>                 Preconditions.checkState(lastResult.ts != Long.MAX_VALUE);
> ```

Lets add a message for the IllegalStateException here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java, line 258 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfKJgHvUBaPh65Thw7:-KwfLCa3YmyZj4KkcXCl:b-ei3vu3) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java#L258)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Lets add a message for the IllegalStateException here.
</blockquote></details>

Out of curiosity why do we need this check?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java, line 258 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfKJgHvUBaPh65Thw7:-KwfLJ1aQGZk1UzddkJt:bmcpoo4) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java#L258)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Out of curiosity why do we need this check?
</blockquote></details>

Will be clearer to have the precondition in continueRow

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java, line 95 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfM7JygssbwkGB4hYB:-KwfM7JygssbwkGB4hYC:b-t0qyi) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java#L95)):*
> ```Java
>         @Override
>         public List<CellTsPairInfo> next() {
>             Preconditions.checkState(hasNext());
> ```

This seems strange, an iterator checking hasNext in next and throwing? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java, line 95 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfM7JygssbwkGB4hYB:-KwfMQxg38bamh0QDi44:btmuwzi) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java#L95)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

This seems strange, an iterator checking hasNext in next and throwing? 
</blockquote></details>

We should either not throw here and return an empty list or just throw the exception (if any) we get when trying to load a page.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java, line 128 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfNWqLsG9R-LBSM2Wn:-KwfNWqLsG9R-LBSM2Wo:b-s8zw0p) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java#L128)):*
> ```Java
> 
>         private AgnosticLightResultSet selectNextPage(ConnectionSupplier conns) {
>             FullQuery query = getQuery();
> ```

query -> fullQuery.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java, line 132 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfNOuOLy6SMMOYve-7:-KwfNOuPDgN52qOny8qP:bcp4x8m) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java#L132)):*
> ```Java
>         }
> 
>         private FullQuery getQuery() {
> ```

Rename to getFullQuery. or we have getQuery().getQuery().

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java, line 182 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfNfEg3-P2uNjX49ip:-KwfNfEg3-P2uNjX49iq:b-o9mpoz) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java#L182)):*
> ```Java
>             } else {
>                 CellTsPairInfo lastResult = Iterables.getLast(results);
>                 Preconditions.checkState(lastResult.ts != Long.MAX_VALUE);
> ```

Lets move the check to continueRow.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
Reviewed 4 of 15 files at r1, 3 of 5 files at r2.
Review status: 5 of 13 files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by gbonik** [comment]
Review status: 4 of 13 files reviewed at latest revision, 12 unresolved discussions.

---

*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java, line 95 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfM7JygssbwkGB4hYB:-KwfRuAorHzmONVx0O_L:b-9lqcqs) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java#L95)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

We should either not throw here and return an empty list or just throw the exception (if any) we get when trying to load a page.
</blockquote></details>

Why is it strange? This is how iterators are supposed to work. Try this: Collections.emptyList().iterator().next();

The only strange this about this is that a more common exception type to throw is NoSuchElementException - I am being lazy here and using Preconditions.checkState instead

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by gbonik** [comment]
Review status: 4 of 13 files reviewed at latest revision, 12 unresolved discussions.

---

*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/DbKvsGetCandidateCellsForSweeping.java, line 40 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwcDeN6e96Tb6r5Feni-r1-40:-KwfTtAyCyYwsgxzioFK:b-hpr29f) ([raw file](https://github.com/palantir/atlasdb/blob/95fb4dd20d513a23f8c00f28dd905a112c73a49a/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/DbKvsGetCandidateCellsForSweeping.java#L40)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Does this Iterator do an eager page load of the kind we were trying to avoid in `OracleCellTsPageLoader.PageIterator`?
</blockquote></details>

Yes, the filtering iterator indeed needs to load the next page when hasNext() is called. However, this is not a problem: every call to hasNext() will be followed by next(), so no work is wasted. The reason why I was trying to avoid that in PageIterators is that CandidateGroupingIterator calls PageIterator.hasNext() after each call to PageIterator.next().

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by gsheasby** [comment]
Review status: 3 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java, line 116 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfIiCxTYqozvxGGwjj:-KwfRQkQ82R5ikpzYB4G:b-3vj6er) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java#L116)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

final?
</blockquote></details>

No - it's updated by this class. This variable now encapsulates all the state.

---

*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java, line 258 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfKJgHvUBaPh65Thw7:-KwfVTdV0kqKTvDO6ZxB:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java#L258)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Will be clearer to have the precondition in continueRow
</blockquote></details>

Done

---

*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java, line 128 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfNWqLsG9R-LBSM2Wn:-KwfVVgiQDp4M5Tl610W:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java#L128)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

query -> fullQuery.
</blockquote></details>

Done

---

*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java, line 132 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfNOuOLy6SMMOYve-7:-KwfVWlGlFqbs0n3R14f:b-j2ym7o) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java#L132)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Rename to getFullQuery. or we have getQuery().getQuery().
</blockquote></details>

Done. `FullQuery` isn't a perfect name (neither is `getQuery()`, but that's beyond the scope of this PR)

---

*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CellTsPairToken.java, line 23 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfIBW0TI_S857oqvx-:-KwfVgpJOcvDE4YSIBr6:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CellTsPairToken.java#L23)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

We'll have a builder that will make the various types of tokens easier to construct..
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by gbonik** [comment]
Review status: 3 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java, line 258 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfKJgHvUBaPh65Thw7:-KwfVBAC32HWCx5JxG5r:b-gu4sr4) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleCellTsPageLoader.java#L258)):*
> Out of curiosity why do we need this check?

We need this check because we increment the timestamp by one to find the next starting position. So if it is equal to Long.MAX_VALUE, it would overflow. This should never be the case, but we check anyway for paranoia reasons

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by gsheasby** [comment]
Review status: 3 of 13 files reviewed at latest revision, 8 unresolved discussions.

---

*[atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java, line 182 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfNfEg3-P2uNjX49ip:-KwfW7UdXt27JwHcO-ui:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/6c3c05a25f3752e8054e0b0667ee3af97cb4c597/atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresCellTsPageLoader.java#L182)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Lets move the check to continueRow.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
*[atlasdb-dbkvs/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CandidateGroupingIteratorTest.java, line 104 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-KwfeGYvbe3tVsznhlGA:-KwfeGYvbe3tVsznhlGB:b-q98xg8) ([raw file](https://github.com/palantir/atlasdb/blob/751f3ae265b64c0b32e94192dde4b10fd7a39723/atlasdb-dbkvs/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CandidateGroupingIteratorTest.java#L104)):*
> ```Java
>                 .isLatestValueEmpty(latestValEmpty)
>                 .numCellsTsPairsExamined(numCellTsPairsExamined)
>                 .sortedTimestamps(ts)
> ```

Should be verify that ts is sorted here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*

**Comment by hsaraogi** [comment]
Some very small comments, looks good overall.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497#-:-Kwfebe5IxNBoL1dhb1F:b40r11w)*

**Comment by hsaraogi** [comment]
Reviewed 9 of 15 files at r1, 2 of 5 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2497)*
